### PR TITLE
[paradoxalarm] Bugfixes and reconnect functionality in case of socket timeout 

### PR DESCRIPTION
**Pull request by theater**
Bugfixes, code improvements and reconnect functionality for the bridge in case of socket timeout

- Implemented reconnect functionality of the bridge handler
- Implemented new parameter in the config of the bridge handler that specifies the reconnect wait time
- Improved and implemented better several toString methods
- Cleaned up logging by reducing several loggings to TRACE from DEBUG
- Several small bugfixes specific to certain null/empty checks here and there

Note: Not sure why AbstractCommunicator and the bridge handler classes appear as completely deleted and remade. Probably I did something wrong with the interactive rebase but no idea what. :(

Signed-off-by: Konstantin Polihronov <polychronov@gmail.com>

### CODE REVIEW

**Comment by Hilbrand** [comment]
> Note: Not sure why AbstractCommunicator and the bridge handler classes appear as completely deleted and remade

Possible windows line ending? Or different character set, and not saved as UTF-8?

**Comment by theater** [comment]
> Possible windows line ending? Or different character set, and not saved as UTF-8?

That was it. I've fixed it but I'm a bit confused since my Eclipse is configured to use unix style. What I did is during solving merge conflicts I accepted the whole file and probably it applied it from another place and not the editor.
Another thing I see is that when I open random other .java files with notepad++ the line endings seem to be windows. I'm very confused...

Thanks for the hint though!  I guess it will ease the review process...

File: bundles/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/communication/AbstractCommunicator.java [hunk 1]
```diff
@@ -115,40 +118,47 @@ protected void sendPacket() {
         SyncQueue syncQueue = SyncQueue.getInstance();
         IRequest request = syncQueue.peekSendQueue();
         try {
-            logger.debug("Sending packet with request={}", request);
+            logger.trace("Sending packet with request={}", request);
             byte[] packetBytes = request.getRequestPayload().getBytes();
             ParadoxUtil.printPacket("Tx Packet:", packetBytes);
             tx.write(packetBytes);
             syncQueue.moveRequest();
+        } catch (SocketException se) {
```
Comment (J-N-K) [inline bundles/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/communication/AbstractCommunicator.java]:
```suggestion
        } catch (SocketException e) {
```

Usually we name all exceptions e

File: bundles/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/communication/AbstractCommunicator.java [hunk 2]
```diff
@@ -115,40 +118,47 @@ protected void sendPacket() {
...
+                logger.trace("Unable to receive proper package for {} time. Rescheduling...", retryCounter);
             } else {
                 IRequest requestInQueue = syncQueue.poll();
-                logger.debug("Error receiving packet after reaching the set treshold of retries. Request: {}",
-                        requestInQueue);
+                logger.debug("Error receiving packet after reaching the set timeout of {}ms. Request: {}",
+                        PACKET_EXPIRATION_TRESHOLD_MILLISECONDS, requestInQueue);
                 retryCounter = 0;
             }
+        } catch (SocketException se) {
```
Comment (J-N-K) [inline bundles/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/communication/AbstractCommunicator.java]:
```suggestion
        } catch (SocketException e) {
```

File: bundles/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/communication/EvoCommunicator.java [hunk 3]
```diff
@@ -77,15 +77,16 @@ protected void receiveRamResponse(IResponse response) throws ParadoxException {
...
             memoryMap.updateElement(ramBlockNumber, parsedResult);
             logger.trace("Result for ramBlock={} is [{}]", ramBlockNumber, parsedResult);
         } else {
             logger.debug("Wrong parsed result. Probably wrong data received in response");
+            return;
         }
 
         // Trigger listeners update when last memory page update is received
-        if (ramBlockNumber == panelType.getRamPagesNumber() && listeners != null && !listeners.isEmpty()) {
+        if (ramBlockNumber == panelType.getRamPagesNumber()) {
```
Comment (J-N-K) [inline bundles/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/communication/EvoCommunicator.java]: Isn't that very risky? Why are you sure that listeners is not null? Otherweise we'll get an NPE https://github.com/openhab/openhab2-addons/blob/34d0f4760e497ba0c092a31381bc440b55f2f885/bundles/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/communication/GenericCommunicator.java#L146

Comment (J-N-K) [reply bundles/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/communication/EvoCommunicator.java]: Forget that. I did not look at the changed implementation in the `GeneriCommunicator`. That would be easier if we had null annotations....

File: bundles/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/communication/ParadoxCommunicatorFactory.java [hunk 4]
```diff
@@ -60,8 +60,8 @@ public IParadoxCommunicator createCommunicator(PanelType panelType)
             throws UnknownHostException, IOException, ParadoxException {
         switch (panelType) {
             case EVO48:
+            case EVO96:
             case EVO192:
-            case EVOHD:
```
Comment (J-N-K) [inline bundles/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/communication/ParadoxCommunicatorFactory.java]: Did you remove the EVOHD intentionally?

File: bundles/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/handlers/ParadoxIP150BridgeHandler.java [hunk 5]
```diff
@@ -131,7 +139,8 @@ private synchronized void doPostOnlineTask(IParadoxInitialLoginCommunicator init
         logger.info("Found {} panel type.", panelType);
         CommunicationState.logout(initialCommunicator);
 
-        // Wait 3 seconds before we create the discovered communicator so we ensure that the socket is closed successfully from
+        // Wait 3 seconds before we create the discovered communicator so we ensure that the socket is closed
+        // successfully from
```
Comment (J-N-K) [inline bundles/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/handlers/ParadoxIP150BridgeHandler.java]:
```suggestion
        // successfully from both ends
```

File: bundles/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/communication/EvoCommunicator.java (hunk 3 above)
Comment (theater) [reply bundles/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/communication/EvoCommunicator.java]: Yep. I moved that inside the method. I think it's better this way...

File: bundles/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/communication/ParadoxCommunicatorFactory.java (hunk 4 above)
Comment (theater) [reply bundles/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/communication/ParadoxCommunicatorFactory.java]: Yes. We figured that the packets used for EVOHD during logon are differeent, so I removed it and I've added EVO96 if I'm not wrong... Basically expectations are that EVO48/96/192 use same protocol. I expected that HD is also using the same but some people tried it and it cannot even pass the logon sequence. Probably if this needs to be supported someone with the necessary skills needs to reverse engineer the protocol and implement new communicator for it.

**Comment by theater** [comment]
I've fixed the proposed and commented the others. Thanks for the review !

**Pull request merged**